### DISCUSSION TITLE
JSON attribute update is not detected if original ORM object was also modified

### DISCUSSION THREAD

**Discussion by guillaumegenthial** [sqlalchemy/sqlalchemy#11004] [Usage Questions] [answered]
### Describe the bug

I want to use a JSON column to store custom fields.

I've run into un-intuitive behavior that caused updates to be skipped (see reproducible steps below on a dumbed-down example).

NOTE: I am aware that JSON columns are not tracked unless using `MutableDict` etc.

A TL;DR of the issue
1. Retrieve an ORM object `x` with a JSON field `x.fields = {"foo": "bar"}`
2. Update a key directly on the ORM object with `x.fields["foo"] = "baz"`
3. Now set the JSON field to a new dictionary with the same value for the updated key  `x.fields = {"foo": "baz"}`
4. No update is triggered

If you skip 2. however, the change is propagated.

I suspect that there is some comparison happening somewhere before flushing to the DB. Since the original object has been updated, it's not able to track the change and incorrectly assumes the DB is up-to-date.

The biggest issue IMO is that this behavior slightly differs from what is expected with literals (setting `x.name = "john"` will always trigger an update).

I'm not sure about how the internals work, but one way to avoid this kind of issue could be to enforce immutability on the JSON fields (since it's not detected anyways). Maybe with a custom json deserializer like so

```python
from immutabledict import immutabledict

def immutable_json_deserializer(data):
    return immutabledict(json.loads(data))


engine = create_engine(
    "sqlite+pysqlite:///:memory:",
    echo=True,
    json_deserializer=immutable_json_deserializer,
)
```

But this feels a bit hacky + is applied session wide.

### Optional link from https://docs.sqlalchemy.org which documents the behavior that is expected

https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON

> Detecting Changes in JSON columns when using the ORM
>
> The [JSON](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON) type, when used with the SQLAlchemy ORM, does not detect in-place mutations to the structure. In order to detect these, the [sqlalchemy.ext.mutable](https://docs.sqlalchemy.org/en/20/orm/extensions/mutable.html#module-sqlalchemy.ext.mutable) extension must be used, most typically using the [MutableDict](https://docs.sqlalchemy.org/en/20/orm/extensions/mutable.html#sqlalchemy.ext.mutable.MutableDict) class. This extension will allow “in-place” changes to the datastructure to produce events which will be detected by the unit of work. See the example at [HSTORE](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.HSTORE) for a simple example involving a dictionary.
> 
> Alternatively, **assigning a JSON structure to an ORM element that replaces the old one will always trigger a change event.**

### SQLAlchemy Version in Use

2.0.25

### DBAPI (i.e. the database driver)

sqlite+pysqlite

### Database Vendor and Major Version

memory

### Python Version

3.11

### Operating system

OSX

### To Reproduce

```python
from uuid import UUID, uuid4

from sqlalchemy import Uuid, create_engine
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    Session,
    mapped_column,
)
from sqlalchemy.types import JSON


class Base(DeclarativeBase):
    """Custom base for DB models."""


class User(Base):
    """User class."""

    __tablename__ = "user"

    uid: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    fields: Mapped[dict] = mapped_column(JSON)


engine = create_engine("sqlite+pysqlite:///:memory:", echo=True)
Base.metadata.create_all(engine)
session = Session(engine)

# Create User
user = User(fields={"foo": "bar"})
session.add(user)
session.commit()

# Force reload
getattr(user, "fields")
user.fields["foo"] = "baz"  # If you comment this line, it works as expected
user.fields = {"foo": "baz"}
session.commit()

assert user.fields["foo"] == "baz"
```

### Error

```
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("user")
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("user")
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-02-12 22:08:11,924 INFO sqlalchemy.engine.Engine
CREATE TABLE user (
	uid CHAR(32) NOT NULL,
	fields JSON NOT NULL,
	PRIMARY KEY (uid)
)


2024-02-12 22:08:11,925 INFO sqlalchemy.engine.Engine [no key 0.00003s] ()
2024-02-12 22:08:11,925 INFO sqlalchemy.engine.Engine COMMIT
2024-02-12 22:08:11,925 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-02-12 22:08:11,926 INFO sqlalchemy.engine.Engine INSERT INTO user (uid, fields) VALUES (?, ?)
2024-02-12 22:08:11,926 INFO sqlalchemy.engine.Engine [generated in 0.00009s] ('8197f5e2586847c39e761f9e7874e153', '{"foo": "bar"}')
2024-02-12 22:08:11,926 INFO sqlalchemy.engine.Engine COMMIT
2024-02-12 22:08:11,926 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine SELECT user.uid AS user_uid, user.fields AS user_fields
FROM user
WHERE user.uid = ?
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine [generated in 0.00005s] ('8197f5e2586847c39e761f9e7874e153',)
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine COMMIT
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine SELECT user.uid AS user_uid, user.fields AS user_fields
FROM user
WHERE user.uid = ?
2024-02-12 22:08:11,927 INFO sqlalchemy.engine.Engine [cached since 0.0005342s ago] ('8197f5e2586847c39e761f9e7874e153',)
Traceback (most recent call last):
  File "test.py", line 41, in <module>
    assert user.fields["foo"] == "baz"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

**Comment by guillaumegenthial** [comment]
I also noticed that the tests expect updates to a non-mutable JSON field to "fail" silently (value can be updated but is not reflected to the DB).

This seems a bit dangerous, I'm curious about the rationale here.

https://github.com/sqlalchemy/sqlalchemy/blob/62b3ca476e879bacb20bb0c520c7c91feca576c0/test/ext/test_mutable.py#L488-L498

**Comment by zzzeek** [comment]
this is expected.  UPDATE does not occur for an attribute that is set to a new value however that value is the same as it was before.     since you are setting an attribute to the identical value that it was before, no UPDATE occurs.

same thing would happen for any other kind of object, like if you had two `Decimal("45.8")` objects.  If I set `obj.x` from one `Decimal("45.8")` to another `Decimal("45.8")`, is that a net change?   people expect this to not be any change.   There's no real way to know the user's intent which was that there was actually a `Decimal("35.12")` there first, then they said `decimal_obj._inner_value = "45.8"` so that it matched.      I dont see any realistic way SQLAlchemy could detect something like this.  if you were to argue, "mutable objects should be tested on identity", I think it's apparent that people serializing new JSON that's equivalent into objects is really common and this would also not go over well.

as you mentioned, if you are mutating JSON objects in place you would use the mutable scalars extension.

**Comment by guillaumegenthial** [comment]
@zzzeek thank you for the quick reply and the explanation. 
The `Decimal` example makes sense, as you say, SQLA cannot realistically detect changes inside all objects.

I still want to "improve" how dict / json are handled to avoid this kind of misuse. 

I identified 2 possibilities
1. propagate mutation tracking inside dictionaries with the mutable scalars extension
2. use immutable dicts so that the original object cannot be updated

The limitation of approach 1. is that the `.get_history()` doesn't track the state change anymore (as expected, see https://github.com/sqlalchemy/sqlalchemy/discussions/7970)

Approach 2. seems to be more robust, with less "magic". So far I used this trick

```python
from immutabledict import immutabledict

def immutable_json_deserializer(data):
    return immutabledict(json.loads(data))


engine = create_engine(
    "sqlite+pysqlite:///:memory:",
    echo=True,
    json_deserializer=immutable_json_deserializer,
)
```

Is there a better way to define immutable dictionaries ? Ideally something like this would be cleaner

```python
class User(Base):
    fields: Mapped[immutabledict] = mapped_column(JSON)
```

**Comment by guillaumegenthial** [comment]
This is what I tried so far

a. Subclass Mutable Dict and trick it (hacky, but could be adapted into an `immutable` ext, similar to the mutable ext?)
```python
from sqlalchemy.ext.mutable import MutableDict

class ImmutableDict(MutableDict):
    def __setitem__(self, key, value):
        raise ValueError()

ImmutableJson = ImmutableDict.as_mutable(JSON)
```

b. Define a custom type (https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator)

```python
from sqlalchemy.types import JSON, TypeDecorator


class ImmutableDict(TypeDecorator):
    """Custom immutable dictionary type using JSON as backend."""

    impl = JSON
    cache_ok = True  # Immutable dict are hashable

    def process_bind_param(self, value, dialect):
        """Convert Python dict to JSON string for storage."""
        if value is not None:
            return json.dumps(value)
        return value

    def process_result_value(self, value, dialect):
        """Convert JSON string from storage to Python dict."""
        if value is not None:
            return immutabledict(json.loads(value))
        return value
```

**Reply by guillaumegenthial** [reply]
I just saw your comment from 2021 where you hint towards using a custom json_deserializer  https://github.com/sqlalchemy/sqlalchemy/discussions/6866#discussioncomment-1259939

**Comment by zzzeek** [comment]
I like the idea of immutable json.  However it should be a complete solution where all dictionaries and sub-dictionaries are immutable and all lists are tuples or some other kind of immutable list.   I did a quick google but didn't find a comprehensive solution, just things like this: https://github.com/Lantero/frozenjsondict .

I think the way to start is to create a custom json deserializer that outputs fully immutable structures, then I'd build a JSON subclass or UserDefinedType that deals in that serializer (it might have to be done as UserDefinedType with a variant to take effect in all cases, e.g. `JSON().with_variant(MyJsonUserDefinedType, 'sqlite', 'postgresql', 'mysql', 'mariadb')` )

**Comment by guillaumegenthial** [comment]
100% agree that a complete solution would be some nested immutable dictionary. I also did a quick google and tried a few things.

1. Subclass `Mapping` and provide a Proxy (avoid data copying). Similar to Raymond Hettinger's suggestion on https://stackoverflow.com/a/9997519. Limitation : is not json serializable (limitation of json...)
2. Subclass `dict` and "disable" `__setitem__`. Limitation: incorrect static typing, only runtime errors

Here is some benchmark (comparing with `immutabledict` library and python's `MappingProxyType` that are NOT nested)

```python
import time
from collections.abc import Mapping
from types import MappingProxyType

from immutabledict import immutabledict


class nestedimmutabledict(dict):
    """Nested Mapping Proxy Type."""

    def __init__(self, data):
        super().__init__(data)
        self._hash = None

    def __getitem__(self, key):
        value = super().__getitem__(key)
        if isinstance(value, Mapping):
            return nestedimmutabledict(value)
        return value

    def __delitem__(self, key):
        raise NotImplementedError()

    def __setitem__(self, key, value):
        raise NotImplementedError()

    def __hash__(self):
        if self._hash is None:
            h = 0
            for key, value in self.items():
                h ^= hash((key, value))
            self._hash = h
        return self._hash


class NestedMappingProxyType(Mapping):
    """Nested Mapping Proxy Type."""

    __slots__ = ("_data", "_hash")

    def __init__(self, data):
        self._data = data
        self._hash = None

    def __getitem__(self, key):
        value = self._data.__getitem__(key)
        if isinstance(value, Mapping):
            return NestedMappingProxyType(value)
        return value

    def __iter__(self):
        return self._data.__iter__()

    def __len__(self):
        return self._data.__len__()

    def __repr__(self):
        return f"{self.__class__.__name__}({self._data.__repr__()})"

    def __hash__(self):
        if self._hash is None:
            h = 0
            for key, value in self.items():
                h ^= hash((key, value))
            self._hash = h
        return self._hash


def benchmark(constructor, num_items, num_repeat):
    times = []
    for _ in range(num_repeat):
        start = time.time()
        items = [
            constructor(
                {
                    "first_name": "John",
                    "last_name": "Doe",
                    "fields": {
                        "age": 42,
                        "street": "Main Street",
                        "city": "New York",
                        "country": "US",
                        "married": False,
                    },
                }
            )
            for _ in range(num_items)
        ]
        end = time.time()
        times.append((end - start) / len(items))
    return sum(times) / num_repeat


if __name__ == "__main__":
    num_items = 10_000
    num_repeat = 100

    # Times are given in e-7 seconds
    # Baseline, time 2.4
    print(benchmark(lambda x: x, num_items=num_items, num_repeat=num_repeat))

    # Non nested, shallow copy, time 6.6 (175% slower)
    print(benchmark(immutabledict, num_items=num_items, num_repeat=num_repeat))

    # Non nested, proxy only, time 2.9 (20% slower)
    print(benchmark(MappingProxyType, num_items=num_items, num_repeat=num_repeat))

    # Nested, proxy only, time 3.9 (60% slower)
    print(benchmark(NestedMappingProxyType, num_items=num_items, num_repeat=num_repeat))

    # Nested, shallow + lazy copy, time 5.2 (116% slower)
    print(benchmark(nestedimmutabledict, num_items=num_items, num_repeat=num_repeat))
```



This could be used with the `json_deserializer` like so (I picked `nestedimmutabledict`  because it works well with json, pydantic, etc.)

```python
from immutabledict import immutabledict

def immutable_json_deserializer(data):
    return nestedimmutabledict(json.loads(data))


engine = create_engine(
    "sqlite+pysqlite:///:memory:",
    echo=True,
    json_deserializer=immutable_json_deserializer,
)
```

**Reply by CaselIT** [reply]
you may also check the sqlalchemy immutabledict from the util package. It does not next though

**Comment by guillaumegenthial** [comment]
FYI I decided to go with another route
1. Annotate ORM classes with `Mapping` instead of dict.
2. Rely on static checks

Pros : no magic, no overhead, clean integration with typing
Cons : no runtime check, requires properly defining types

So in practice

```python
from typing import Mapping
from uuid import UUID, uuid4

from sqlalchemy import Uuid, create_engine
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    Session,
    mapped_column,
)
from sqlalchemy.types import JSON


class Base(DeclarativeBase):
    """Custom base for DB models."""


class User(Base):
    """User class."""

    __tablename__ = "user"

    uid: Mapped[UUID] = mapped_column(Uuid, primary_key=True, default=uuid4)
    fields: Mapped[Mapping] = mapped_column(JSON)


engine = create_engine("sqlite+pysqlite:///:memory:", echo=True)
Base.metadata.create_all(engine)
session = Session(engine)

# Create User
user = User(fields={"foo": "bar"})
session.add(user)
session.commit()

# Force reload
getattr(user, "fields")
user.fields["foo"] = "baz"
session.commit()

assert user.fields["foo"] == "baz"
```

which gives when linting

```txt
>> mypy test4.py
test.py:39: error: Unsupported target for indexed assignment ("Mapping[Any, Any]")  [index]
Found 1 error in 1 file (checked 1 source file)
```

**Reply by guillaumegenthial** [reply]
Also, if I know the dict can be nested I annotate with `Mapping[Any, Mapping | Any]` for example (it works, not sure if there is something better from a typing semantic's point of view)

**Reply by zzzeek** [reply, marked as answer]
i think that's a great solution for now, after all this is what typing is for

**Reply by guillaumegenthial** [reply]
Thanks for the help @zzzeek !
